README.md
Keight.rb README
================

($Release: 0.0.0 $)


Overview
--------

Keight.rb is the very fast web application framework for Ruby.
It runs about 100 times faster than Rails, and 20 times faster than Sinatra.

*Keight.rb is under development and is subject to change without notice.*


Benchmarks
----------

Measured with `app.call(env)` style in order to exclude server overhead:

| Framework | Request            | usec/req | req/sec  |
|:----------|:-------------------|---------:|---------:|
| Rails     | GET /api/hello     |    738.7 |   1353.7 |
| Rails     | GET /api/hello/123 |    782.2 |   1278.4 |
| Sinatra   | GET /api/hello     |    144.1 |   6938.3 |
| Sinatra   | GET /api/hello/123 |    158.4 |   6313.8 |
| Rack      | GET /api/hello     |      9.9 | 101050.9 |
| Keight    | GET /api/hello     |      7.6 | 132432.8 |
| Keight    | GET /api/hello/123 |     10.6 |  94705.9 |

* Ruby     2.2.3
* Rails    4.2.4
* Sinatra  1.4.6
* Rack     1.6.4 (= Rack::Request + Rack::Response)
* Keight   0.1.0

(Script: https://gist.github.com/kwatch/c634e91d2d6a6c4b1d40 )


Quick Tutorial
--------------

```console
$ ruby -v     # required Ruby >= 2.0
ruby 2.2.3p173 (2015-08-18 revision 51636) [x86_64-darwin14]
$ mkdir gems
$ export GEM_HOME=$PWD/gems
$ export PATH=$GEM_HOME/bin:$PATH

$ gem install keight
$ vi hello.rb      # see below
$ vi config.ru     # != 'config.rb'
$ rackup -p 4423 config.ru
```

hello.rb:

```ruby
# -*- coding: utf-8 -*-
require 'keight'

class HelloAction < K8::Action

  mapping ''       , :GET=>:do_index, :POST=>:do_create
  mapping '/{id}'  , :GET=>:do_show, :PUT=>:do_update, :DELETE=>:do_delete

  def do_index
    return {"message"=>"Hello"}   # JSON
    #return "<h1>Hello</h1>"      # HTML
  end

  def do_show(id)
    ## 'id' or 'xxx_id' will be converted into integer.
    return {"id"=>id}
    #return "<h1>id=#{id.inspect}</h1>"
  end

  def do_create    ; "<p>create</p>"; end
  def do_update(id); "<p>update</p>"; end
  def do_delete(id); "<p>delete</p>"; end

end
```

config.ru:

```ruby
# -*- coding: utf-8 -*-
require 'keight'

mapping = [
  ['/api', [
    ['/hello'    , "./hello:HelloAction"],
    ### or
    #['/hello'     , HelloAction],
  ]],
]
app = K8::RackApplication.new(mapping)

run app
```

Open http://localhost:4423/api/hello or http://localhost:4423/api/hello/123
with your browser.

How do you like it? Try the following steps to generate your project.

```console
$ mkdir gems                         # if necessary
$ export GEM_HOME=$PWD/gems          # if necessary
$ export PATH=$GEM_HOME/bin:$PATH    # if necessary

$ gem install keight    # or: gem install boilerpl8
$ k8rb project myapp1   # or: boilerpl8 github:kwatch/keight-ruby myapp1
## select CSS framework
**    1. None
**    2. Bootstrap
**    3. Pure (recommended)
** Which CSS framework do you like? [1-3]: 1

$ cd myapp1
$ export APP_MODE=dev                # 'dev', 'prod', or 'stg'
$ rake -T
$ ls public
$ rake server port=4423
$ open http://localhost:4423/
$ ab -n 10000 -c 100 http://localhost:4423/api/hello.json
```


Command `k8rb`
--------------

Keight.rb provices `k8rb`.

* `k8rb project myapp1` creates new project.  
  (Note: This is equvarent to `boilerpl8 github:kwatch/keight-ruby myapp1`.)

* `k8rb cdnjs -d static/lib jquery 3.1.0` downloads jQuery files
  from cdnjs.com and stores into `static/lib` directory.  
  (Note: This is equivarent to `cdnget cdnjs jquery 3.1.0 static/lib`.)

`k8rb` command is provided mainly for backward compatibility.
You can use [boilerpl8](https://github.com/kwatch/boilerpl8/tree/ruby)
and [cdnget](https://github.com/kwatch/cdnget/tree/ruby-release)
inead of `k8rb project` and `k8rb cdnjs`.
`k8rb` is specific to Keight.rb, but both boilerpl8 and cdnget are
available in any project.


CheatSheet
----------

```ruby
require 'keight'

class HelloAction < K8::Action

  ## mapping
  mapping ''             , :GET=>:do_hello_world
  mapping '/{name:str}'  , :GET=>:do_hello

  ## request, response, and helpers

  def do_hello_world
    do_hello('World')
  end

  def do_hello(name)

    ## request
    @req                   # K8::Request object (!= Rack::Request)
    @req.env               # Rack environment
    @req.meth              # ex: :GET, :POST, :PUT, ...
    @req.request_method    # ex: "GET", "POST", "PUT", ...
    @req.path              # ex: '/api/hello.json'
    @req.path_ext          # ex: '.json'
    @req.query_string      # query string (String)
    @req.query             # query string (Hash)
    @req.form              # form data (Hash)
    @req.multipart         # multipart form data ([Hash, Hash])
    @req.json              # JSON data (Hash)
    @req.params            # query or form (not multipart nor json!)
    @req.cookies           # cookies (Hash)
    @req.xhr?              # true when requested by jQuery etc
    @req.client_ip_addr    # ex: '127.0.0.1'

    ## response
    @resp                  # K8::Response object (!= Rack::Response)
    @resp.status_code      # ex: 200
    @resp.status_code=(i)  # ex: i=200
    @resp.status_line      # ex: "200 OK"
    @resp.status           # alias of @resp.status_code
    @resp.status=(i)       # alias of @resp.status_code=(i)
    @resp.headers          # Hash object
    @resp.set_cookie(k, v) # cookie
    @resp.content_type        # same as @resp.headers['Content-Type']
    @resp.content_type=(s)    # same as @resp.headers['Content-Type'] = s
    @resp.content_length      # same as @resp.headers['Content-Length']
    @resp.content_length=(i)  # same as @resp.headers['Content-Length'] = i.to_s

    ## session (requires Rack::Session)
    @sess[key] = val       # set session data
    @sess[key]             # get session data

    ## helpers
    token = csrf_token()   # get csrf token
    validation_failed()    # same as @resp.status = 422
    return redirect_to(location, 302, flash: "message")
    return send_file(filepath)

  end


  ## hook methods

  def before_action        # setup
    super
  end

  def after_action(ex)     # teardown
    super
  end

  def handle_content(content)   # convert content
    if content.is_a?(Hash)
      @resp.content_type = 'application/json'
      return JSON.dump(content)
    end
    super
  end

  def handle_exception(ex)   # exception handler
    proc_ = WHEN_RAISED[ex.class]
    return self.instance_exec(ex, &proc_) if proc_
    super
  end

  WHEN_RAISED = {}
  WHEN_RAISED[NotFound]     = proc {|ex| ... }
  WHEN_RAISED[NotPermitted] = proc {|ex| ... }

  def csrf_protection_required?
    x = @req.method
    return x == :POST || x == :PUT || x == :DELETE || x == :PATCH
  end

end

## urlpath mapping
urlpath_mapping = [
  ['/'               , './app/page/welcome:WelcomePage'],
  ['/api', [
    ['/books'        , './app/api/books:BooksAPI'],
    ['/books/{book_id}/comments'
                     , './app/api/books:BookCommentsAPI'],
    ['/orders'       , './app/api/orders:OrdersAPI'],
  ]],
]

## application
opts = {
  urlpath_cache_size:  0,    # 0 means cache disabled
}
app = K8::RackApplication.new(urlpath_mapping, opts)

## misc
p HelloAction[:do_update].meth          #=> :PUT
p HelloAction[:do_update].path(123)     #=> "/api/books/123"
p HelloAction[:do_update].form_action_attr(123)
                                        #=> "/api/books/123?_method=PUT"
```


URL Mapping
-----------

Rails-like mapping:

```ruby
class BookAPI < K8::Action

  mapping ''       , :GET=>:do_index, :POST=>:do_create
  mapping '/new'   , :GET=>:do_new
  mapping '/{id}'  , :GET=>:do_show, :PUT=>:do_update, :DELETE=>:do_delete
  mapping '/{id}/edit', :GET=>:do_edit

  def do_index()   ; {"list": []}; end
  def do_create()  ; {"list": []}; end
  def do_new()     ; {"item": id}; end
  def do_show(id)  ; {"item": id}; end
  def do_edit(id)  ; {"item": id}; end
  def do_update(id); {"item": id}; end
  def do_delete(id); {"item": id}; end
  # (Note: 'do_' prefix is NOT mandatory.)

end
```

Data type and pattern:

```ruby
class BookAPI < K8::Action

  ##
  ## data type and pattern
  ##
  mapping '/{name:str<[^/]+>}' , :GET=>:do_show1
  mapping '/{id:int<\d+>}'     , :GET=>:do_show2
  mapping '/{birthday:date<\d\d\d\d-\d\d-\d\d>}', :GET=>:do_show3

  ##
  ## default pattern
  ##  - '/{name:str}' is same as '/{name:str<[^/]+>}'
  ##  - '/{id:int}' is same as '/{id:int<\d+>}'
  ##  - '/{birthday:date}' is same as '/{birthday:date<\d\d\d\d-\d\d-\d\d>}'
  ##
  mapping '/{name:str}'        , :GET=>:do_show1
  mapping '/{id:int}'          , :GET=>:do_show2
  mapping '/{birthday:date}'   , :GET=>:do_show3

  ##
  ## default data type
  ##  - 'id' and '*_id' are int type
  ##  - 'date' and '*_date' are date type
  ##  - others are str type
  ##
  mapping '/{name}'      , :GET=>:do_show1   # same as '{name:str}'
  mapping '/{id}'        , :GET=>:do_show2   # same as '{id:int}'
  mapping '/{birth_date}', :GET=>:do_show3   # same as '{birth_date:date}'

  ##
  ## pattern with default data type
  ##
  mapping '/{code:<[A-Z]{3}>}', ...  # same as '/{code:str<[A-Z]{3}>'
  mapping '/{id:<[1-9]\d*>}'  , ...  # same as '/{id:int<\d{4}>}'

end
```

Path extension (such as '.json' or '.html'):

```ruby
class FooAPI < K8::Action

  ## ex: '.*' is same as '{_:<(?:\w+)?>}' which matches to any extension.
  mapping '.*'       , :GET=>:do_index

  ## ex: '/{id}.*' is same as '/{id}{_:<(?:\w+)?>}' which matches to any extension.
  mapping '/{id}.*'  , :GET=>:do_show

  def do_index
    p @req.path_ext    #=> ex: '.json', '.html', and so on
  end

  def do_show(id)
    p @req.path_ext    #=> ex: '.json', '.html', and so on
  end

end
```

URL mapping helper:

```ruby
## for example
p BookAPI[:do_index].meth        #=> :GET
p BookAPI[:do_index].path        #=> "/api/books"
p BookAPI[:do_create].meth       #=> :POST
p BookAPI[:do_create].path       #=> "/api/books"
p BookAPI[:do_show].meth         #=> :GET
p BookAPI[:do_show].path(123)    #=> "/api/books/123"
p BookAPI[:do_update].meth       #=> :PUT
p BookAPI[:do_update].path(123)  #=> "/api/books/123"
p BookAPI[:do_delete].meth       #=> :DELETE
p BookAPI[:do_delete].path(123)  #=> "/api/books/123"

p BookAPI[:do_index ].form_action_attr()   #=> "/api/books"
p BookAPI[:do_create].form_action_attr()   #=> "/api/books"
p BookAPI[:do_show  ].form_action_attr(9)  #=> "/api/books/9"
p BookAPI[:do_update].form_action_attr(9)  #=> "/api/books/9?_method=PUT"
p BookAPI[:do_delete].form_action_attr(9)  #=> "/api/books/9?_method=DELETE"
```

Show URL mappings:

```console
$ k8rb project myapp1   # or: boilerpl8 github:kwatch/keight-ruby myapp1
$ cd myapp1
$ rake mapping:text       # list url mapping
$ rake mapping:yaml       # list in YAML format
$ rake mapping:json       # list in JSON format
$ rake mapping:jquery     # list for jQuery
$ rake mapping:angularjs  # list for AngularJS
```


Topics
------


### Make Routing More Faster


CHANGES.md
Changes
=======

Release 0.3.0 (2016-10-13)
--------------------------

### Enhancements

* URL path parameter format can accept parameter type (`int`, `str`, `date`).  
  ex: `/api/books/{id:int}`, `/blog/{entry_date:date}`

* `K8::Util::TemporaryFile` class added. It allows you to create temporary
  file which will be removed automatically *after* sending response body.  
  See document of `K8::Util::TemporaryFile` for details.

* `K8::Util::ShellCommand` class added. It allows you to send command output
  as response body. For example, execute SQL, convert into CSV, and send it
  as response.  
  See document of `K8::Util::ShellCommand` for details.

* `@action_args` is available in `K8::Action` class.  
  `@action_args` represents URL path parameter arguments.  
  For example, `@action_args == [123]` when urlpath pattern is `/books/{id}`
  and request path is `/books/123`.

* `@req.params_form()`, `@req.params_multipart()` and `@req.params_json()`
  now take max content length.  
  ex:

      strs = @req.params_form(10*1024*104)  # max: 10MB
      strs = @req.params_form               # same as above

      strs, files = @req.params_multipart(100*1024*1024)  # max: 100MB
      strs, files = @req.params_multipart                 # same as above

      hash = @req.params_json(10*1024*104)  # max: 10MB
      hash = @req.params_json               # same as above

* Add `K8::ActionMapping#urlpath_rexp`.
  You can confirm regexp of action mapping by that attribute.  
  ex:

      ## what regexp is used for request routing?
      mappings = [
        ['/api', [
          ['/books',    BooksAction],
          ['/authors',  AuthorsAction],
        ]
      ]
      p K8::ActionMapping.new(mapping).urlpath_rexp

* Regard '.*' at end of urlpath pattern as extension pattern.  
  ex:

      class BooksAPI < K8::Action

        ## ex: '.*' is same as '{_:<(?:\.\w+)?>}'
        mapping '.*'     , :GET=>:do_index

        ## ex: '/{id}.*' is same as '/{id}{_:<(?:\.\w+)?>}'
        mapping '/{id}.*', :GET=>:do_show

        def do_index
          p @req.path_ext    #=> ex: '.json', '.html', etc
        end
        def do_show(id)
          p @req.path_ext    #=> ex: '.json', '.html', etc
        end
      end

* Define `@req.path_ext` which returns such as `.json` or `.html`.

* Define `@resp.status_line` which returns such as `"200 OK"` or `"302 Found"`.

* `@resp.set_cookie()` now accepts Time object as `expires` keyword arg.

* New HTTP response status codes are added.  
  #

      103 Checkpoint                           # Unofficial
      306 Switch Proxy                         # -
      308 Permanent Redirect                   # RFC 7538
      421 Misdirected Request                  # RFC 7540
      428 Precondition Required                # RFC 6585
      429 Too Many Requests                    # RFC 6585
      431 Request Header Fields Too Large      # RFC 6585
      444 No Response                          # nginx
      451 Unavailable For Legal Reasons        # -
      495 SSL Certificate Error                # nginx
      496 SSL Certificate Required             # nginx
      497 HTTP Request Sent to HTTPS Port      # nginx
      499 Client Closed Request                # nginx
      509 Bandwidth Limit Exceeded             # Apache Web Server/cPanel
      511 Network Authentication Required      # RFC 6585
      520 Unknown Error                        # CloudFlare
      521 Web Server Is Down                   # CloudFlare
      522 Connection Timed Out                 # CloudFlare
      523 Origin Is Unreachable                # CloudFlare
      524 A Timeout Occurred"                  # CloudFlare
      525 SSL Handshake Failed                 # CloudFlare
      526 Invalid SSL Certificate              # CloudFlare

* `k8rb cdnjs` now calls CDNJS.com api instead of scraping HTML page.


### Incomppatible changes

* URL path parameter format is changed.  
  (old) `/books/{id:\d+}`  
  (new) `/books/{id:int}` or `/books/{id:int<\d+>}`  
  See document for details.

* Rename `@req.method` to `@req.meth`.  
  `@req.method` is also available, but not recommended. Use `@req.meth`.

* Rename `K8::BaseAction[]#method` to `K8::BaseAction[]#meth`.  
  ex:

      p BooksAction[:do_show].meth     #=> :GET
      p BooksAction[:do_update].meth   #=> :PUT
      p BooksAction[:do_delete].meth   #=> :DELETE

* Rename `K8::BaseAction[]#urlpath()` to `BooksAction[]#path()`.  
  ex:

      p BooksAction[:do_show].path(123)     #=> "/api/books/123"
      p BooksAction[:do_update].path(123)   #=> "/api/books/123"
      p BooksAction[:do_delete].path(123)   #=> "/api/books/123"

* Rename `@current_action` to `@action_name` in `K8::Action` class.
  It represents current action name, such as `:do_index` or `:do_show`.

* `@req.params` now raises `K8::PayloadParseError` for JSON data,
  because `@req.params` should return hash object of string, but
  JSON data contains non-string values.
  Use `@req.json` instead fo `@req.params`.

* `@req.params` now raises `K8::PayloadParseError` for multipart data,
  because `@req.params` should return hash object of string, but
  multipart form contains file data which is not a string value.
  Use `@req.multipart` instead fo `@req.params`.

* Rename `K8::Request` class to `K8::RackRequest`.

* Rename `K8::Response` class to `K8::RackResponse`.


bench/bench.rb
# -*- coding: utf-8 -*-

$LOAD_PATH << '.'

def _version(cmdopt)
  if cmdopt == '0'
    return nil
  end
  begin
    return yield
  rescue LoadError
    if cmdopt.to_s.empty?
      return nil
    else
      raise
    end
  end
end

version_rack = _version($rack) { require 'rack'            ; Rack.release }
version_sina = _version($sina) { require 'sinatra/base'    ; Sinatra::VERSION }
version_mplx = _version($mplx) { require 'rack/multiplexer'; Rack::Multiplexer::VERSION }
version_k8   = _version($k8  ) { require 'keight'          ; K8::RELEASE }
version_jet  = _version($jet ) { require 'rack/jet_router' ; Rack::JetRouter::RELEASE }

$api_entries = ('a'..'z').each_with_index.map {|x, i| "%s%02d" % [x*3, i+1] }


if version_rack

  class RackApp1
    def call(env)
      [200, {"Content-Type"=>"text/html"}, ["<h1>hello</h1>"]]
    end
  end

  class RackApp2
    def call(env)
      req  = Rack::Request.new(env)
      [200, {"Content-Type"=>"text/html"}, ["<h1>hello</h1>"]]
    end
  end

  class RackApp3
    def call(env)
      resp = Rack::Response.new
      #[resp.status, resp.headers, ["<h1>hello</h1>"]]
      [200, {"Content-Type"=>"text/html"}, ["<h1>hello</h1>"]]
    end
  end

  class RackApp4
    def call(env)
      req  = Rack::Request.new(env)
      resp = Rack::Response.new
      #[resp.status, resp.headers, ["<h1>hello</h1>"]]
      [200, {"Content-Type"=>"text/html"}, ["<h1>hello</h1>"]]
    end
  end

  rack_app1 = RackApp1.new
  rack_app2 = RackApp2.new
  rack_app3 = RackApp3.new
  rack_app4 = RackApp4.new

end


if version_sina

  class SinaApp < Sinatra::Base

    set :sessions   , false
    set :logging    , false
    set :protection , false
    set :x_cascade  , false

    for x in $api_entries
      get    "/api/#{x}"          do "<h1>index</h1>"  end
      post   "/api/#{x}"          do "<h1>create</h1>" end
      #get    "/api/#{x}/new"      do "<h1>new</h1>"    end
      get    "/api/#{x}/:id"      do "<h1>show</h1>"   end
      put    "/api/#{x}/:id"      do "<h1>update</h1>" end
      delete "/api/#{x}/:id"      do "<h1>delete</h1>" end
      #get    "/api/#{x}/:id/edit" do "<h1>edit</h1>"   end
    end

  end

  sina_app = SinaApp.new

end


if version_mplx

  mplx_app = proc {
    #
    proc_ = proc {|env|
      [200, {"Content-Type"=>"text/html"}, ["<h1>hello</h1>"]]
      #req  = Rack::Request.new(env)
      #resp = Rack::Response.new
      #[resp.status, resp.headers, ["<h1>index</h1>"]]
    }
    #
    app = Rack::Multiplexer.new
    for x in $api_entries
      app.get    "/api/#{x}",           proc_
      app.post   "/api/#{x}",           proc_
      #app.get    "/api/#{x}/new",       proc_
      app.get    "/api/#{x}/:id",       proc_
      app.post   "/api/#{x}/:id",       proc_
      app.delete "/api/#{x}/:id",       proc_
      #app.get    "/api/#{x}/:id/edit",  proc_
    end
    #
    app
  }.call()

end


if version_k8

  class DummyAction < K8::Action
    mapping '',       :GET=>:do_index, :POST=>:do_create
    #mapping '/new',   :GET=>:do_new
    mapping '/{id}',  :GET=>:do_show, :PUT=>:do_edit, :DELETE=>:do_delete
    #mapping '/{id}/edit', :GET=>:do_edit
    def do_index        ; "<h1>index</h1>"; end
    def do_create       ; "<h1>create</h1>"; end
    def do_new          ; "<h1>new</h1>"; end
    def do_show(id)     ; "<h1>show</h1>"; end
    def do_update(id)   ; "<h1>update</h1>"; end
    def do_delete(id)   ; "<h1>delete</h1>"; end
    def do_edit(id)     ; "<h1>edit</h1>"; end
  end
  #
  k8_app = proc {
    mapping = [
      ['/api', $api_entries.map {|x| ["/#{x}", DummyAction] }],
    ]
    opts = {
      urlpath_cache_size: ($k8size || 0).to_i,
      enable_urlpath_param_range: $k8range != '0',
    }
    K8::RackApplication.new(mapping, opts)
  }.call()

end


if version_jet

  jet_app = proc {
    #
    jet_proc = proc {|env|
      [200, {"Content-Type"=>"text/html"}, ["<h1>hello</h1>"]]
      #req  = Rack::Request.new(env)
      #resp = Rack::Response.new
      #[resp.status, resp.headers, ["<h1>index</h1>"]]
    }
    methods1 = {:GET=>jet_proc, :POST=>jet_proc}
    methods2 = {:GET=>jet_proc, :PUT=>jet_proc, :DELETE=>jet_proc}
    #
    arr = $api_entries.map {|x| ["/#{x}", [['', methods1], ['/:id', methods2]]] }
    jet_mapping = [
        ['/api', $api_entries.map {|x| ["/#{x}", [
                                           ['',     methods1],
                                           ['/:id', methods2],
                                       ]] }
        ],
    ]
    #
    jet_opts = {
      urlpath_cache_size: ($jetcache || 0).to_i,
      enable_urlpath_param_range: $jetrange != '0',
    }
    #
    Rack::JetRouter.new(jet_mapping, jet_opts)
  }.call()

end


def _chk(tuple)
  tuple[0] == 200  or raise "200 expected but got #{tuple[0]}"
end

require 'rack' unless defined?(Rack)
$environ = Rack::MockRequest.env_for("http://localhost/", method: 'GET')

def newenv(path)
  env = $environ.dup
  env['PATH_INFO'] = path
  env
end


require 'benchmarker'

N = ($N || 100000).to_i

Benchmarker.new(:width=>30, :loop=>N) do |bm|

  urlpaths = [
    '/api/aaa01',
    '/api/aaa01/123',
    '/api/zzz26',
    '/api/zzz26/789',
  ]

  tuple = nil

  puts ""
  puts "** rack            : #{version_rack || '-'}"
  puts "** rack-jet_router : #{version_jet  || '-'}"
  puts "** rack-multiplexer: #{version_mplx || '-'}"
  puts "** sinatra         : #{version_sina || '-'}"
  puts "** keight          : #{version_k8   || '-'}"
  puts ""
  puts "** N=#{N}"

  ### empty task

  bm.empty_task do
    newenv("/api/hello")
  end


  ### Rack
  if version_rack
    rack_labels = ['Rack', 'R:Rq', 'R:Rs', 'RqRs']
    rack_apps = [rack_app1, rack_app2, rack_app3, rack_app4]
    for rack_app, label in rack_apps.zip(rack_labels)
      bm.task("(#{label}) /some/where") do
        tuple = rack_app.call(newenv("/some/where"))
      end
      _chk(tuple)
    end
  end


  ### Sinatra
  if version_sina
    for upath in urlpaths
      bm.task("(Sina) #{upath}") do
        tuple = sina_app.call(newenv(upath))
      end
      _chk(tuple)
    end
  end


  ### Rack::Multiplexer
  if version_mplx
    for upath in urlpaths
      bm.task("(Mplx) #{upath}") do
        tuple = mplx_app.call(newenv(upath))
      end
      _chk(tuple)
    end
  end


  ### Keight

  if version_k8
    for upath in urlpaths
      bm.task("(K8  ) #{upath}") do
        tuple = k8_app.call(newenv(upath))
      end
      _chk(tuple)
    end
  end


  ### Rack::JetRouter

  if version_jet
    for upath in urlpaths
      bm.task("(JetR) #{upath}") do
        tuple = jet_app.call(newenv(upath))
      end
      _chk(tuple)
    end
  end

end


__END__

###
### Example result
###

$ ruby -I../lib -I. -s bench.rb -N=1000000

benchmarker.rb:   release 0.0.0
RUBY_VERSION:     2.3.0
RUBY_PATCHLEVEL:  0
RUBY_PLATFORM:    x86_64-darwin15

** rack            : 1.6.4
** rack-jet_router : 1.1.0
** rack-multiplexer: 0.0.8
** sinatra         : 1.4.6
** keight          : 0.2.0

** N=1000000

##                                  user       sys     total      real
(Empty)                           7.3000    0.0900    7.3900    7.4131
(Rack) /some/where                0.8400   -0.0400    0.8000    0.7903
(R:Rq) /some/where                0.8400   -0.0500    0.7900    0.7743
(R:Rs) /some/where                6.1600   -0.0500    6.1100    6.0898
(RqRs) /some/where                6.4200   -0.0600    6.3600    6.3407
(Sina) /api/aaa01                84.5700   18.0800  102.6500  102.7069
(Sina) /api/aaa01/123            94.1300   18.4800  112.6100  112.6892
(Sina) /api/zzz26               125.7300   19.0800  144.8100  145.0454
(Sina) /api/zzz26/789           133.8000   18.9600  152.7600  152.9070
(Mplx) /api/aaa01                 5.0200    0.1700    5.1900    5.1821
(Mplx) /api/aaa01/123            16.7100    0.3100   17.0200   17.0086
(Mplx) /api/zzz26                23.3700   -0.0400   23.3300   23.3190
(Mplx) /api/zzz26/789            35.6500    0.0100   35.6600   35.6661
(K8  ) /api/aaa01                 5.9200    0.0300    5.9500    5.9423
(K8  ) /api/aaa01/123             9.2200    0.0900    9.3100    9.2934
(K8  ) /api/zzz26                 6.2700    0.0600    6.3300    6.3158
(K8  ) /api/zzz26/789            10.0300    0.1300   10.1600   10.1547
(JetR) /api/aaa01                 1.1800    0.0400    1.2200    1.2154
(JetR) /api/aaa01/123             5.0400    0.1100    5.1500    5.1431
(JetR) /api/zzz26                 0.9600    0.0300    0.9900    0.9681
(JetR) /api/zzz26/789             5.6900    0.1100    5.8000    5.7946

## Ranking                          real
(R:Rq) /some/where                0.7743 (100.0%) ********************
(Rack) /some/where                0.7903 ( 98.0%) ********************
(JetR) /api/zzz26                 0.9681 ( 80.0%) ****************
(JetR) /api/aaa01                 1.2154 ( 63.7%) *************
(JetR) /api/aaa01/123             5.1431 ( 15.1%) ***
(Mplx) /api/aaa01                 5.1821 ( 14.9%) ***
(JetR) /api/zzz26/789             5.7946 ( 13.4%) ***
(K8  ) /api/aaa01                 5.9423 ( 13.0%) ***
(R:Rs) /some/where                6.0898 ( 12.7%) ***
(K8  ) /api/zzz26                 6.3158 ( 12.3%) **
(RqRs) /some/where                6.3407 ( 12.2%) **
(K8  ) /api/aaa01/123             9.2934 (  8.3%) **
(K8  ) /api/zzz26/789            10.1547 (  7.6%) **
(Mplx) /api/aaa01/123            17.0086 (  4.6%) *
(Mplx) /api/zzz26                23.3190 (  3.3%) *
(Mplx) /api/zzz26/789            35.6661 (  2.2%)
(Sina) /api/aaa01               102.7069 (  0.8%)
(Sina) /api/aaa01/123           112.6892 (  0.7%)
(Sina) /api/zzz26               145.0454 (  0.5%)
(Sina) /api/zzz26/789           152.9070 (  0.5%)

## Matrix                           real     [01]     [02]     [03]     [04]     [05]     [06]     [07]     [08]     [09]     [10]     [11]     [12]     [13]     [14]     [15]     [16]     [17]     [18]     [19]     [20]
[01] (Rack2) /some/where          0.7743   100.0%   102.1%   125.0%   157.0%   664.2%   669.2%   748.3%   767.4%   786.5%   815.6%   818.9%  1200.2%  1311.4%  2196.6%  3011.5%  4606.1% 13264.0% 14553.1% 18731.7% 19747.0%
[02] (Rack1) /some/where          0.7903    98.0%   100.0%   122.5%   153.8%   650.7%   655.7%   733.2%   751.9%   770.5%   799.1%   802.3%  1175.9%  1284.8%  2152.1%  2950.5%  4512.7% 12995.2% 14258.3% 18352.2% 19346.9%
[03] (JetR) /api/zzz26            0.9681    80.0%    81.6%   100.0%   125.5%   531.2%   535.3%   598.5%   613.8%   629.0%   652.4%   654.9%   959.9%  1048.9%  1756.9%  2408.7%  3684.0% 10608.8% 11639.9% 14982.1% 15794.1%
[04] (JetR) /api/aaa01            1.2154    63.7%    65.0%    79.7%   100.0%   423.2%   426.4%   476.8%   488.9%   501.0%   519.6%   521.7%   764.6%   835.5%  1399.4%  1918.6%  2934.5%  8450.4%  9271.7% 11933.8% 12580.6%
[05] (JetR) /api/aaa01/123        5.1431    15.1%    15.4%    18.8%    23.6%   100.0%   100.8%   112.7%   115.5%   118.4%   122.8%   123.3%   180.7%   197.4%   330.7%   453.4%   693.5%  1997.0%  2191.1%  2820.2%  2973.1%
[06] (Mplx) /api/aaa01            5.1821    14.9%    15.3%    18.7%    23.5%    99.2%   100.0%   111.8%   114.7%   117.5%   121.9%   122.4%   179.3%   196.0%   328.2%   450.0%   688.3%  1982.0%  2174.6%  2799.0%  2950.7%
[07] (JetR) /api/zzz26/789        5.7946    13.4%    13.6%    16.7%    21.0%    88.8%    89.4%   100.0%   102.5%   105.1%   109.0%   109.4%   160.4%   175.2%   293.5%   402.4%   615.5%  1772.5%  1944.7%  2503.1%  2638.8%
[08] (K8) /api/aaa01              5.9423    13.0%    13.3%    16.3%    20.5%    86.6%    87.2%    97.5%   100.0%   102.5%   106.3%   106.7%   156.4%   170.9%   286.2%   392.4%   600.2%  1728.4%  1896.4%  2440.9%  2573.2%
[09] (Rack3) /some/where          6.0898    12.7%    13.0%    15.9%    20.0%    84.5%    85.1%    95.2%    97.6%   100.0%   103.7%   104.1%   152.6%   166.7%   279.3%   382.9%   585.7%  1686.5%  1850.5%  2381.8%  2510.9%
[10] (K8) /api/zzz26              6.3158    12.3%    12.5%    15.3%    19.2%    81.4%    82.0%    91.7%    94.1%    96.4%   100.0%   100.4%   147.1%   160.8%   269.3%   369.2%   564.7%  1626.2%  1784.3%  2296.6%  2421.0%
[11] (Rack4) /some/where          6.3407    12.2%    12.5%    15.3%    19.2%    81.1%    81.7%    91.4%    93.7%    96.0%    99.6%   100.0%   146.6%   160.2%   268.2%   367.8%   562.5%  1619.8%  1777.2%  2287.5%  2411.5%
[12] (K8) /api/aaa01/123          9.2934     8.3%     8.5%    10.4%    13.1%    55.3%    55.8%    62.4%    63.9%    65.5%    68.0%    68.2%   100.0%   109.3%   183.0%   250.9%   383.8%  1105.2%  1212.6%  1560.7%  1645.3%
[13] (K8) /api/zzz26/789         10.1547     7.6%     7.8%     9.5%    12.0%    50.6%    51.0%    57.1%    58.5%    60.0%    62.2%    62.4%    91.5%   100.0%   167.5%   229.6%   351.2%  1011.4%  1109.7%  1428.4%  1505.8%
[14] (Mplx) /api/aaa01/123       17.0086     4.6%     4.6%     5.7%     7.1%    30.2%    30.5%    34.1%    34.9%    35.8%    37.1%    37.3%    54.6%    59.7%   100.0%   137.1%   209.7%   603.9%   662.5%   852.8%   899.0%
[15] (Mplx) /api/zzz26           23.3190     3.3%     3.4%     4.2%     5.2%    22.1%    22.2%    24.8%    25.5%    26.1%    27.1%    27.2%    39.9%    43.5%    72.9%   100.0%   152.9%   440.4%   483.3%   622.0%   655.7%
[16] (Mplx) /api/zzz26/789       35.6661     2.2%     2.2%     2.7%     3.4%    14.4%    14.5%    16.2%    16.7%    17.1%    17.7%    17.8%    26.1%    28.5%    47.7%    65.4%   100.0%   288.0%   316.0%   406.7%   428.7%
[17] (Sina) /api/aaa01          102.7069     0.8%     0.8%     0.9%     1.2%     5.0%     5.0%     5.6%     5.8%     5.9%     6.1%     6.2%     9.0%     9.9%    16.6%    22.7%    34.7%   100.0%   109.7%   141.2%   148.9%
[18] (Sina) /api/aaa01/123      112.6892     0.7%     0.7%     0.9%     1.1%     4.6%     4.6%     5.1%     5.3%     5.4%     5.6%     5.6%     8.2%     9.0%    15.1%    20.7%    31.6%    91.1%   100.0%   128.7%   135.7%
[19] (Sina) /api/zzz26          145.0454     0.5%     0.5%     0.7%     0.8%     3.5%     3.6%     4.0%     4.1%     4.2%     4.4%     4.4%     6.4%     7.0%    11.7%    16.1%    24.6%    70.8%    77.7%   100.0%   105.4%
[20] (Sina) /api/zzz26/789      152.9070     0.5%     0.5%     0.6%     0.8%     3.4%     3.4%     3.8%     3.9%     4.0%     4.1%     4.1%     6.1%     6.6%    11.1%    15.3%    23.3%    67.2%    73.7%    94.9%   100.0%
